### ISSUE TITLE
Rework of ElasticsearchWriter performance data

### ISSUE THREAD

**Issue by martialblog** [Icinga/icinga2#10511]
Hi, 

I have some suggestions for improving the ElasticsearchWriter.

Currently the ElasticsearchWriter creates a new field for every metric in a given check command.
This leads to a vast number of fields in the Elasticsearch/OpenSearch indices.

Here are some examples:

```json
{
  "_index": "icinga",
  "_id": "UQdo5ZcByMLqDokf4UFI",
  "_source": {
    "@timestamp": "2025-07-07T15:02:20.573+0000",
    "check_command": "load",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_load",
      "-c",
      "10,6,4",
      "-w",
      "5,4,3"
    ],
    "check_result.execution_end": "2025-07-07T15:02:20.573+0000",
    "check_result.execution_start": "2025-07-07T15:02:20.570+0000",
    "check_result.execution_time": 0.0027740001678466797,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0003490447998046875,
    "check_result.output": "OK - load average: 0.24, 0.25, 0.26",
    "check_result.perfdata.load1.crit": 10,
    "check_result.perfdata.load1.min": 0,
    "check_result.perfdata.load1.value": 0.24,
    "check_result.perfdata.load1.warn": 5,
    "check_result.perfdata.load15.crit": 4,
    "check_result.perfdata.load15.min": 0,
    "check_result.perfdata.load15.value": 0.26,
    "check_result.perfdata.load15.warn": 3,
    "check_result.perfdata.load5.crit": 6,
    "check_result.perfdata.load5.min": 0,
    "check_result.perfdata.load5.value": 0.25,
    "check_result.perfdata.load5.warn": 4,
    "check_result.schedule_end": "2025-07-07T15:02:20.573+0000",
    "check_result.schedule_start": "2025-07-07T15:02:20.569+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "homestead",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "load",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-07T15:02:20.573+0000",
    "type": "icinga2.event.checkresult"
  }
},
{
  "_index": "icinga",
  "_id": "-j6O6pcBNgAT8LEo7v53",
  "_score": 1,
  "_source": {
    "@timestamp": "2025-07-08T15:01:56.548+0000",
    "check_command": "ping4",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_ping",
      "-4",
      "-H",
      "127.0.0.1",
      "-c",
      "200,15%",
      "-w",
      "100,5%"
    ],
    "check_result.execution_end": "2025-07-08T15:01:56.548+0000",
    "check_result.execution_start": "2025-07-08T15:01:52.466+0000",
    "check_result.execution_time": 4.081341981887817,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0006992816925048828,
    "check_result.output": "PING OK - Packet loss = 0%, RTA = 0.07 ms",
    "check_result.perfdata.pl.crit": 15,
    "check_result.perfdata.pl.min": 0,
    "check_result.perfdata.pl.unit": "percent",
    "check_result.perfdata.pl.value": 0,
    "check_result.perfdata.pl.warn": 5,
    "check_result.perfdata.rta.crit": 0.2,
    "check_result.perfdata.rta.min": 0,
    "check_result.perfdata.rta.unit": "seconds",
    "check_result.perfdata.rta.value": 7.4e-05,
    "check_result.perfdata.rta.warn": 0.1,
    "check_result.schedule_end": "2025-07-08T15:01:56.548+0000",
    "check_result.schedule_start": "2025-07-08T15:01:52.466+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "Demo1",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "ping4",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-08T15:01:56.548+0000",
    "type": "icinga2.event.checkresult"
  }
}
```

This means, the ElasticsearchWriter is practically unusable in large Icinga2 setups. Because a high amount of fields can lead to performance degradations and memory issues in clusters (See https://www.elastic.co/guide/en/elasticsearch/reference/8.18/mapping-settings-limit.html )

Elasticsearch/OpenSearch only allow 1000 fields as default, to safeguard clusters against indices that create too many fields.

As mentioned, the Icinga2 ElasticsearchWriter creates a new field for each metric in a check plugin. Each new plugin, each new metric adds new fields to the index. See also: https://github.com/Icinga/icinga2/issues/6805

# A possible solution

One solution could be to redesign the output of the ElasticsearchWriter to use an array of nested objects for the performance data.
These are supported in both Elasticsearch and OpenSearch:

- https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/nested
- https://docs.opensearch.org/docs/latest/field-types/supported-field-types/nested/

> The nested type is a specialised version of the object data type that allows arrays of objects to be indexed in a way that they can be queried independently of each other.

Example from the docs:

```
{
  "patients": [
    {"name" : "John Doe", "age" : 56, "smoker" : true},
    {"name" : "Mary Major", "age" : 85, "smoker" : false}
  ]
}
```
There are some pitfalls to look out for. By default, arrays of objects are flattened during indexing. Meaning the ElasticsearchWrite needs to make sure the nested field is used.

An example of how this would look like in Icinga2:

```json
{
  "_index": "icinga",
  "_id": "UQdo5ZcByMLqDokf4UFI",
  "_score": 1,
  "_source": {
    "@timestamp": "2025-07-07T15:02:20.573+0000",
    "check_command": "load",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_load",
      "-c",
      "10,6,4",
      "-w",
      "5,4,3"
    ],
    "check_result.execution_end": "2025-07-07T15:02:20.573+0000",
    "check_result.execution_start": "2025-07-07T15:02:20.570+0000",
    "check_result.execution_time": 0.0027740001678466797,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0003490447998046875,
    "check_result.output": "OK - load average: 0.24, 0.25, 0.26",
    "check_result.perfdata": [
        {
            "metric_name": "load1",
            "crit": 10,
            "min": 0,
            "value": 0.24,
            "warn": 5,
        },
        {
            "metric_name": "load15",
            "crit": 4,
            "min": 0,
            "value": 0.26,
            "warn": 3,
        },
        {
            "metric_name": "load5",
            "crit": 6,
            "min": 0,
            "value": 0.25,
            "warn": 4,
        },
    ]
    "check_result.schedule_end": "2025-07-07T15:02:20.573+0000",
    "check_result.schedule_start": "2025-07-07T15:02:20.569+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "homestead",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "load",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-07T15:02:20.573+0000",
    "type": "icinga2.event.checkresult"
  }
}
```

Benefits:

1. Avoids creating new fields for each metric in a check plugin
2. Make the index fields more predictable when searching
3. Makes the perfdata independently searchable

Obviously, this would be a breaking change. 

However, I don't see how the ElasticsearchWriter would ever be usable practice without reworking the current performance data output, regardless of how it is reworked. Since every cluster is doomed to run into the field mapping limit eventually with the current implementation, as fields cannot be removed from the mapping.

Regards,
Markus

**Comment by martialblog** [comment]
Apparently, this is also how [icingabeat](https://github.com/Icinga/icingabeat/blob/d667e10439873a9082ba58e979c7bba86439c977/beater/statuspoller.go#L83) writes the performance data:

```go
case interface{}:
  key = "perfdata." + perfdata.(map[string]interface{})["label"].(string)
  value = perfdata
  event.Fields.Put(target_key+key, value)
```

**Comment by martialblog** [comment]
Though about this some more.  Another possible solution - however less elegant - could be to simply provide the "raw" perfdata, similar to the PerfdataWriter:

```
"check_result.perfdata_raw": "rta=0.080000ms;100.000000;200.000000;0.000000 pl=0%;5;15;0;"
```

That would provide a stable field name and avoid the field mapping explosion.
However, users would have to parse that data in order to work with it, that's not great.

Maybe there could also be an additional field with the data already split:

```
"check_result.perfdata": [
  "rta=0.080000ms;100.000000;200.000000;0.000000",
  "pl=0%;5;15;0;"
]

"check_result.perfdata_raw": "time=0.010865s;;;0.000000;10.000000"
"check_result.perfdata": [
  "time=0.010865s;;;0.000000;10.000000"
]

"check_result.perfdata_raw": "load1=0.170;5.000;10.000;0; load5=0.260;4.000;6.000;0; load15=0.270;3.000;4.000;0;"
"check_result.perfdata": [
  "load1=0.170;5.000;10.000;0;",
  "load5=0.260;4.000;6.000;0;",
  "load15=0.270;3.000;4.000;0;"
]
```

Better but still, not great.

People would still have to write their own parsing to get to the heart of the data.

**Comment by martialblog** [comment]
I did some testing with the PerfdataWriter so that the data is transformed into the format I suggested. Example from the Elasticsearch index:



```json
{
  "_index": "perfdata",
  "_id": "F7fPN5gBBKdijTHgnNMN",
  "_source": {
    "@timestamp": "2025-07-23T15:03:29.422432908Z",
    "performance_data": [
      {
        "value": 0.73,
        "warn": 5,
        "crit": 10,
        "min": 0,
        "label": "load1"
      },
      {
        "value": 0.32,
        "warn": 4,
        "crit": 6,
        "min": 0,
        "label": "load5"
      },
      {
        "value": 0.12,
        "warn": 3,
        "crit": 4,
        "min": 0,
        "label": "load15"
      }
    ],
  }
}
```

Now I can do this to search for the performance data:

```bash
curl -X POST "http://localhost:9200/perfdata/_search" -H 'Content-Type: application/json' -d'
  {
    "query": {
        "bool": {
         "must": [
          { "match": { "performance_data.label": "load1" }}
        ]
      }
     }
  }'
```

This is still not optimal, since the a search for a given label will return all documents with this label. However, this is more workable than the current implementation.

The index still uses dynamic mapping - just like the ElasticsearchWriter would - and this is the resulting mapping:

```json
"performance_data": {
  "properties": {
    "crit": {
      "type": "float"
    },
    "label": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      }
    },
    "max": {
      "type": "float"
    },
    "min": {
      "type": "float"
    },
    "unit": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      }
    },
    "value": {
      "type": "float"
    },
    "warn": {
      "type": "float"
    }
  }
}
```

So no more mapping explosion.

**Comment by martialblog** [comment]
Closing this. See the discussion here: https://github.com/Icinga/icinga2/pull/10518#issuecomment-3953106726

> Our strategy is to focus on the OTLPMetricsWriter (https://github.com/Icinga/icinga2/pull/10685) and deprecate the existing ElasticsearchWriter in favor it (https://github.com/Icinga/icinga2/issues/10730).
